### ISSUE TITLE
Denoise Filter Library for Sensors

### ISSUE THREAD

**Issue by lblim** [solettaproject/soletta#2310]
#### Task Description
Introduce denoise filter library into Soletta, it is needed for noisy sensors such as gyroscope and magnetometer
Type of denoise filters
- Mean / moving average filter
- Median filter

The filters and FBP node should be applied to all supported sensor types regardless 
Physical connection of sensors (either through sensor hub or direct connect)
Logical connection of sensors  (either through IIO or non-IIO)

**Comment by lblim** [comment]
@laykuanloon @fulong82 @elvinongbl

**Comment by barbieri** [comment]
did you check the buffer types? I recall @glima wrote them to address these

**Comment by glima** [comment]
Yes,  if it's just mean/median you want, we already have support for that at least for integer types. Float could be added easily as well.

**Comment by lblim** [comment]
@barbieri and @glima Thanks for pointing this out. Are you referring to https://github.com/solettaproject/soletta/blob/master/src/modules/flow/int/int.c

@laykuanloon Please take a look at the source code to see if it meets our requirements, we may extend on top of it if we need more than that. Also to expand for other data type such as float.

**Comment by glima** [comment]
* LB Lim <notifications@github.com> [2016-09-20 18:13:08 -0700]:

> @barbieri and @glima Thanks for pointing this out. Are you referring to https://github.com/solettaproject/soletta/blob/master/src/modules/flow/int/int.c

Yes, exactly (int/buffer node).

>
> @laykuanloon Please take a look at the source code to see if it meets our requirements, we may extend on top of it if we need more than that. Also to expand for other data type such as float.
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/solettaproject/soletta/issues/2310#issuecomment-248482794

--
Gustavo Lima Chaves
Intel - Open Source Technology Center

**Comment by barbieri** [comment]
@lblim did that work for you?

**Comment by barbieri** [comment]
@laykuanloon ping

**Comment by laykuanloon** [comment]
@barbieri 
I implement denoise inside my sample app. [https://github.com/solettaproject/soletta/commit/45117b9b9951204bbb2c8c7d6709f76d0e0d4a30](url)

Is better to have a denoise that can work for C and sol_direction_vector data type. What do you think?

**Comment by barbieri** [comment]
@lblim I agree, I'll propose this as a project, thanks.

In this project we will also propose denoise based on redundancy, like done in drones, one can use another sensor at different sampling rate, driver and mount, then use them to provide the same information and use the redundancy to reduce noise.

In addition to traditional buffer + function filtering as done before.
